subsys/bluetooth/mesh/foundation.h

```diff
@@ -154,17 +154,43 @@ void bt_mesh_attention(struct bt_mesh_model *model, uint8_t time);
 
 #include <zephyr/sys/byteorder.h>
 
-static inline void key_idx_pack(struct net_buf_simple *buf,
-				uint16_t idx1, uint16_t idx2)
+static inline void key_idx_pack_pair(struct net_buf_simple *buf, uint16_t idx1, uint16_t idx2)
 {
 	net_buf_simple_add_le16(buf, idx1 | ((idx2 & 0x00f) << 12));
 	net_buf_simple_add_u8(buf, idx2 >> 4);
 }
 
-static inline void key_idx_unpack(struct net_buf_simple *buf,
-				  uint16_t *idx1, uint16_t *idx2)
+static inline void key_idx_pack_list(struct net_buf_simple *buf, uint16_t *arr, size_t len)
+{
+	size_t i;
+
+	for (i = 0; i < len - 1; i += 2) {
+		key_idx_pack_pair(buf, arr[i], arr[i + 1]);
+	}
+
+	if (i < len) {
+		net_buf_simple_add_le16(buf, arr[i]);
+	}
+}
+
+static inline void key_idx_unpack_pair(struct net_buf_simple *buf, uint16_t *idx1, uint16_t *idx2)
 {
 	*idx1 = sys_get_le16(&buf->data[0]) & 0xfff;
 	*idx2 = sys_get_le16(&buf->data[1]) >> 4;
 	net_buf_simple_pull(buf, 3);
 }
+
+static inline int key_idx_unpack_list(struct net_buf_simple *buf, uint16_t *dst_arr, size_t dst_len)
+{
+	size_t i;
+
+	for (i = 0; (i + 1) < dst_len && buf->len >= 3; i += 2) {
+		key_idx_unpack_pair(buf, &dst_arr[i], &dst_arr[i + 1]);
+	}
+
+	if (i < dst_len && buf->len >= 2) {
+		dst_arr[i++] = net_buf_simple_pull_le16(buf) & 0xfff;
+	}
+
+	return buf->len > 0 ? -EMSGSIZE : i;
```

Comment:
> Additionally, we could extend this function to provide amount of keys in the buf if dst_val is NULL.

hmm... im looking into this. It's a bit tricky since there are several illigal values for the buffer size, like 1, 5 , 7 etc. Im not quite sure how to handle this if we have an unexpected number of bytes in the buffer.